Prepare CHERI-C support

fbrausse opened the pull request:
This PR prepares for support the CHERI-C semantics, which will be part of another PR. Essentially, this PR does:

- add support for CMake to recognize and build against [CHERI-Clang](https://github.com/CTSRD-CHERI/llvm-project), which is a fork of LLVM intended to support the new semantics for MIPS64 and RISC-V CHERI-enabled architectures. There is another [fork](https://git.morello-project.org/morello/llvm-project/-/tree/morello/dev/) by ARM that supports Aarch64 targets. These compilers add a few new API constructs such as `clang::BuiltinType::UIntCap` and `clang::CK_PointerToCHERICapability` which this PR handles - behind #ifdefs depending on whether ESBMC is built against CHERI-Clang or the regular one.
- extend the `configt` and the embedded `ansi_ct` to:
  - know about whether ESBMC is supposed to verify standard C or CHERI-C,
  - what mode of CHERI-C: `purecap` or `hybrid` (this is set by the new cmd-line option `--cheri (hybrid|purecap)`, default: off),
  - whether or not compression of CHERI capabilities is enabled (cmd-line flag `--cheri-uncompressed`, default: off)
  - derive the size of capability-like types from these settings (in CHERI-C they are usually twice the size of regular pointers), i.e. there now is a distinction between `pointer_width()` and `capability_width()`.
- fix various places throughout the code base that previously assumed pointer-related sizes to correspond to `word_size` or `int_width` or similar. Actually, there is no correspondence between any C types and `word_size`. Such assumption should be removed throughout: `sizeof(void *)` and `sizeof(size_t)` are no longer related.
- introduce new `typet` for the new types `[u]intcap_t` (the CHERI-capability equivalent for `[u]intptr_t`)
- add a flag `can_carry_provenance` to irep1 types that corresponds to the similarly named flag from CHERI-Clang; this flag basically says that operations on expressions of such types should copy the associated validity tag.
- add a new set of pre-compiled GOTO libraries for the the CHERI hybrid and purecap modes (now we have 12),
- passes a bunch of new flags to the Clang frontend when the CHERI support is enabled (these unfortunately vary between platforms, see `clang_c_language.cpp` for details)
- add a new pseudo expression `pointer_capability2t` that, applied to capability-typed sub-expression, should yield the capability metadata in symbolic form; this functionality is not used in this PR and will be extended in the next one
- add support for the new metadata to the `smt_memspace`, though this is not complete, yet

fbrausse committed:
[cheri] add support for morello-clang-11.0.0 (w/o libc due to <fenv.h> bug)

fbrausse committed:
[smt] attempt at plugging the free'd object address space hole

fbrausse committed:
[smt] move from generic pointer_type() to ptraddr_type()

fbrausse committed:
[smt] cheri: fix for adding field pointer_cap_info

fbrausse committed:
[config clang-c-frontend] cheri: enable purecap arch/flags only for --cheri purecap

fbrausse committed:
[regression] cheri-c: remove unnecessary include-dir + pass --cheri hybrid

fbrausse committed:
[elibc] typo

fbrausse committed:
[cheri c2goto/library] fix

fbrausse committed:
[smt] cheri: hide "pointer_cap_info" behind config.cheri

fbrausse committed:
cheri: WIP

fbrausse committed:
[cheri] prepare support for RISC-V CHERI target

fbrausse committed:
[cheri cmake] conditionally enable CHERI-support

fbrausse committed:
[regression] add cheri-128

fbrausse committed:
[cheri] more regression tests

fbrausse committed:
[cheri] init

lucasccordeiro committed:
[cheri] added the cap_ptr operational model [tests]

fbrausse committed:
[cheri util/config] pointer_width() == capability_width() only in purecap mode

fbrausse committed:
[irep2] fix for pointer-arith: result has ptrdiff_t type of address-width

fbrausse committed:
cheri: WIP

fbrausse committed:
[cheri] translate __builtin_cheri_length_get

fbrausse committed:
[cheri] WIP

fbrausse committed:
[smt] add comments explaining the fragile nature of the previous fix

fbrausse committed:
more ptr-int-type fixes

fbrausse committed:
[cheri] WIP

fbrausse committed:
clang-format

fbrausse committed:
add --nostdinc option to esbmc, c2goto, config and clang-c-frontend

fbrausse committed:
[cheri] add irep types for [u]intcap

fbrausse committed:
[regression] cheri-c: remove commented #include

fbrausse committed:
[cheri] add --cheri + generate corresponding internal clibs

fbrausse committed:
[irep2] add pointer_capability2t expression type akin pointer_(object|offset)2t

fbrausse committed:
[elic] cheri: pass __ESBMC_CHERI_MORELLO__ #define to support Morello compressed-cap

fbrausse committed:
[clang-c-frontend] pass down option "-cheri" + put compiler include-dir at end of list

fbrausse committed:
[cheri] WIP

fbrausse committed:
[cmake] define default compiler opts also for clang

fbrausse committed:
clang-format

fbrausse committed:
[cheri] WIP: intro ansi_c.capability_width() + --cheri {hybrid,purecap} + --cheri-uncompressed

fbrausse committed:
[cheri] streamline sysroot option processing

fbrausse committed:
[smt-memspace] NULL object is never alive, yet address constraints are required

fbrausse committed:
[smt-memspace] better variable naming

fbrausse committed:
[util] c_typecast: appearently also code expressions fail

fbrausse committed:
[util] config: only print "overriding flavor" warning if flavor is set

fbrausse committed:
[readme] add ISSTA information directly at the top (cheri-clang branch only)

fbrausse committed:
[smt] add field "pointer_cap_info" to pointer_struct

fbrausse committed:
[regression] add cases from #775

fbrausse committed:
[smt] casts: assert CHERI-concentrate format for this implementation

fbrausse committed:
[clang-c-frontend] c_typecast: handle [u]intcap_t like [un]signedbv

fbrausse committed:
cheri: fix one more case of pointer_offset not being of signed addr-width

fbrausse committed:
[clang-c-frontend] bundle clang headers into a single cheaders.c file

fbrausse committed:
[cheri] default to mips64c128-unknown-linux for Clang + build cheri-enabled goto libs

fbrausse committed:
[cheri] update build instructions to refer to cheribuild

fbrausse committed:
[util] optionst: add another get_option() overload

The existing variants of get_option() discard the existence of the requested
option in the map, while the new overload only conditionally sets its "value"
parameter. This is helpful in cases a hard-coded default exists which can be
overridden by options.

fbrausse committed:
clang-format

fbrausse committed:
[cheri] add external lib cheri-compressed-cap to impl cheri_length_get

fbrausse committed:
[cheri] fix div2t signedness issue + collect some regressions tests

fbrausse committed:
[cheri] move Morello fenv.c hack to add_bundled_library_sources()

fbrausse committed:
[readme] update video link

fbrausse committed:
[cheri] update build instructions

olivierpierre committed:
[cheri] Add clear tag regression test

fbrausse committed:
[smt] casts: take capability info into account when converting to/from pointers

Fixes in regression/esbmc/:
- regression/esbmc/github_212
- regression/esbmc/github_426_2
- regression/esbmc/github_562

fbrausse committed:
[cheri] migrate [u]intcap_typet

fbrausse committed:
[c2goto/library] remove #include-s <assert.h> and <stdint.h> + only enable cheri-functions for 128-bit mode

fbrausse committed:
[cheri] WIP

fbrausse committed:
[clang-c-frontend] format cheri intrinsics

fbrausse committed:
[regression] add cheri-128/00_types-(hybrid|purecap)

fbrausse committed:
make __ESBMC_alloc_size of type size_t[]

fbrausse committed:
[util/config cheri] default to mips64el on --cheri for now: big-endian bit-fields are b0rked #571

fbrausse committed:
cheri: fixes for rebasing

fbrausse committed:
c_link: fix logging observed in regression/nonz3/overflow_12_adpcm

fbrausse committed:
[clang-c-adjust] avoid converting constant array exprs to &arr[0] due to non-matching size().type()

string2array was inventing a hard-coded size() expressions for the resulting
string expression. This was signedbv and mismatched the unsignedbv array size
from the frontend. Down the line, this caused gen_typecast() to convert a
declaration of the form

  char str[] = "constant"; // size() is (size_t)9

into something like

  char str[] = &"constant"[0]; // size() is (int)9

Later on in smt-conv, an FAM special case zeroed out the RHS causing problems
with the regression/cstd/memmove family of regression tests.

fbrausse committed:
[regression] move esbmc/github_839 to esmbc-old/ since it uses --old-frontend

fbrausse committed:
linter...

fbrausse committed:
[cmake] add support for FetchContent of Catch2 v3.x

Summary: add v3.x support for Catch2 as the cheri-compressed-cap library wants
that one.

FetchContent is utterly broken [1]. As usual, it implements its own "dependency
management", which consists of the, quote,

> "first to declare, wins" approach

Not only does that break everything importing another package as then that
project could potentially override other packages down the line, no, it also
introduces even more state into the CMake files in form of sequence of
declarations. However, FetchContent is not the only part of the build system
that relies on the specific order of declarations. This results in projects
using FetchContent packages to solve the nice NP complete dependency
management problem manually in their CMakeLists.txt every time a dependency is
updated...

[1] <https://gitlab.kitware.com/cmake/cmake/-/issues/22904>

fbrausse commented:
Does anyone know how to run cmake on the Github CI with the version given in the README from the logs?
> Runner Image
  Image: ubuntu-22.04
  Version: 20230217.1
  Included Software: https://github.com/actions/runner-images/blob/ubuntu22/20230217.1/images/linux/Ubuntu2204-Readme.md

It claims cmake to be version 3.25.2, however the build log [says](https://github.com/esbmc/esbmc/actions/runs/4232319408/jobs/7351907462#step:7:8)
>  CMake Error at CMakeLists.txt:4 (cmake_minimum_required):
  CMake 3.18 or higher is required.  You are running version 3.16.3

lucasccordeiro commented on regression/cheri-128/00_types-hybrid/test.desc:
```diff
@@ -0,0 +1,4 @@
+CORE
+main.c
+--cheri hybrid --64 --big-endian
+^VERIFICATION SUCCESSFUL$
```
Can you also write the failed version of this regression test?

lucasccordeiro commented on regression/cheri-128/00_types-purecap/test.desc:
```diff
@@ -0,0 +1,4 @@
+CORE
+main.c
+--cheri purecap --64 --big-endian
+^VERIFICATION SUCCESSFUL$
```
Can you also write the failed version of this regression test?

lucasccordeiro commented on regression/cheri-128/01_cheri_ptr11_failed-hybrid/test.desc:
```diff
@@ -0,0 +1,4 @@
+CORE
+main.c
+--cheri hybrid --big-endian
+^VERIFICATION FAILED$
```
Do we have the successful version of this regression test?

lucasccordeiro commented on regression/cheri-128/02_temporal-purecap/test.desc:
```diff
@@ -0,0 +1,4 @@
+CORE
+main.c
+--cheri purecap --big-endian --force-malloc-success
+^VERIFICATION FAILED$
```
Do we have the successful version of this regression test?

lucasccordeiro commented on src/c2goto/library/builtin_libs.c:
```diff
@@ -6,3 +6,268 @@ int __ESBMC_sync_fetch_and_add(int *ptr, int value)
+  };
+};
+
+__SIZE_TYPE__ __esbmc_cheri_base_get(void *__capability cap)
```
Shall we add a short documentation about this method? You could use the doxygen style here.

lucasccordeiro commented on src/c2goto/library/builtin_libs.c:
```diff
@@ -6,3 +6,268 @@ int __ESBMC_sync_fetch_and_add(int *ptr, int value)
+  __PTRADDR_TYPE__ top = cheri_gettop(cap);
+#endif
+  __ESBMC_assert(cheri_gettag(cap), "tag-violation c2exception");
+  // __ESBMC_assert(!cc128_is_cap_sealed(&comp) /*cheri_getsealed(cap)*/, "seal-violation c2exception");
```
Do you know if we need to keep this comment code here?

fbrausse replied:
Not needed, removed.

lucasccordeiro commented on src/c2goto/library/builtin_libs.c:
```diff
@@ -6,3 +6,268 @@ int __ESBMC_sync_fetch_and_add(int *ptr, int value)
+  // __ESBMC_assume(ret.cursor == cursor);
+  // __ESBMC_assume(ret.pesbt == comp.cr_pesbt);
+  /*
+  __ESBMC_assume(cheri_gettag(ret.cap) == cheri_gettag(cap));
```
Do you know if we need to keep this comment code here?

fbrausse replied:
Removed.

lucasccordeiro commented on src/c2goto/library/builtin_libs.c:
```diff
@@ -6,3 +6,268 @@ int __ESBMC_sync_fetch_and_add(int *ptr, int value)
+__SIZE_TYPE__ __esbmc_cheri_length_get(void *__capability cap)
+{
+#if 1
+  /*
```
Do you know if we need to keep this comment code here?

fbrausse replied:
Removed.

lucasccordeiro commented on src/clang-c-frontend/clang_c_convert.cpp:
```diff
@@ -2477,6 +2517,16 @@ bool clang_c_convertert::get_cast_expr(
+    /* An explicit __cheri_tocap means this value might be tagged. */
+  case clang::CK_CHERICapabilityToPointer:
+    /* both should not be generated in purecap mode */
+    // assert(0);
```
Do you know if we need to keep this comment code here?

fbrausse replied:
Removed.

lucasccordeiro commented on src/clang-c-frontend/clang_c_convert.cpp:
```diff
@@ -2477,6 +2517,16 @@ bool clang_c_convertert::get_cast_expr(
     break;
+
+#ifdef ESBMC_CHERI_CLANG
+  case clang::CK_PointerToCHERICapability:
```
What is the purpose of these cases?

fbrausse replied:
The purpose is to track whether these constants change their name in upstream CHERI-Clang and to remind me to translate these conversions properly in a later PR.

lucasccordeiro commented on src/clang-c-frontend/clang_c_language.cpp:
```diff
@@ -29,6 +29,13 @@ clang_c_languaget::clang_c_languaget()
   // Build the compile arguments
   build_compiler_args(clang_headers_path());
+
+#if 0
```
Do you know if we need to keep this comment code here?

fbrausse replied:
The code is useful for debugging the clang invocation params. I've put it behind Verbosity::Debug.

lucasccordeiro commented on src/irep2/irep2_expr.h:
```diff
@@ -2401,6 +2403,10 @@ class mul2t : public mul_expr_methods
   mul2t(const type2tc &type, const expr2tc &v1, const expr2tc &v2)
     : mul_expr_methods(type, mul_id, v1, v2)
   {
+#if 0
```
We should perform this checking for the entire irep2! :-)

fbrausse replied:
I agree, but would suggest to do this in another PR. These assertion still trigger way too often for me to fix here.

lucasccordeiro commented on src/irep2/irep2_type.cpp:
```diff
@@ -175,7 +175,7 @@ unsigned int vector_type2t::get_width() const
 unsigned int pointer_type2t::get_width() const
 {
-  return config.ansi_c.pointer_width;
+  return config.ansi_c.pointer_width(); /* CHERI-TODO */
```
What is missing here?

fbrausse replied:
I've clarified the TODO. We basically need to take into account whether this is a pointer or a capability (i.e. `can_carry_provenance`).

lucasccordeiro commented on src/solvers/smt/smt_casts.cpp:
```diff
@@ -345,6 +345,60 @@ smt_astt smt_convt::convert_typecast_to_ints_from_bool(const typecast2t &cast)
+
+static expr2tc capability_struct_from_cap(const expr2tc &cap)
+{
+  // assert(is_pointer_type(cap));
```
Do you know if we need to keep this comment code here?

fbrausse replied:
I've changed it to a CHERI-TODO comment for me.

lucasccordeiro commented on src/solvers/smt/smt_conv.cpp:
```diff
@@ -718,6 +728,19 @@ smt_astt smt_convt::convert_ast(const expr2tc &expr)
+  {
+    assert(config.ansi_c.cheri);
+    const pointer_capability2t &obj = to_pointer_capability2t(expr);
+    // Potentially walk through some typecasts
```
Why?

fbrausse replied:
For the same reason the cases
- expr2t::pointer_offset_id, and
- expr2t::pointer_object_id

do. :) Basically, the expression can contain typecasts which under these kinds of expressions are ignored.

fbrausse commented:
I've updated this PR's description. While I sort through @lucasccordeiro's comments, please note that enabling this functionality requires a different Clang to be compiled against. I'm working on providing the CI with one.

lucasccordeiro commented:
@fbrausse: Can I ask you whether you could solve the conflicts and rebase this PR with the master branch?

fbrausse committed:
[dereference, smt] add some more support for dynamic array sizes

Fixes #1099.

fbrausse committed:
replace usage of type_byte_size() + potential handling of dyn-sized exc.

fbrausse committed:
[solvers] pointer-logic: use type matching array size's for indexing

fbrausse committed:
ensure width of type of index into array matches that of its size

Fixes #1091

fbrausse committed:
[regression] add case from #1067

fbrausse commented:
Blocked on #1003 due to CMake version.

lucasccordeiro commented:
@fbrausse: can I ask you to run your PR over the SV-COMP benchmarks?

lucasccordeiro commented:
@fbrausse: can I ask you whether this issue is still blocked?

fbrausse committed:
[cmake] support more of Catch2-v3

fbrausse commented:
SV-COMP with 30s. Master:
```
Statistics:          26459 Files
  correct:           14374
    correct true:     8161
    correct false:    6213
  incorrect:            60
    incorrect true:     30
    incorrect false:    30
  unknown:           12025
  Score:             21095 (max: 43043)
```
This PR (de70b7e7b8aa453f2921ee669f3a06d207cecffe):
```
Statistics:          26459 Files                                                                         
  correct:           14418                                                                               
    correct true:     8189                                                                               
    correct false:    6229                                                                               
  incorrect:           186                                                                               
    incorrect true:     28                                                                               
    incorrect false:   158                                                                               
  unknown:           11855                                                                               
  Score:             19183 (max: 43043)
```
I'll take a look at the (many) new incorrect false. Though the other stats have gotten better.

fbrausse committed:
choose large enough type for type_byte_size_bits*()

Fixes error with e.g. SV-COMP benchmark c/array-memsafety/array01-alloca-2.i.

fbrausse commented:
With a0215cde504f8c5e50e2c8437254e43371fff9fa applied, this PR does better now:
```
Statistics:          26459 Files
  correct:           14406
    correct true:     8178
    correct false:    6228
  incorrect:            63
    incorrect true:     28
    incorrect false:    35
  unknown:           11990
  Score:             21128 (max: 43043)
```
New incorrect are in unreach-call:
- 2x previously timeouted
- 3x previously unknown (segfault), all fail with an assertion in a debug build, all are #584 for 64-bit mode

fbrausse commented:
Previously wrong are now:
- 1x timeout
- 1x correct, unexpectedly this is `c/memsafety/cmp-freed-ptr.i` from #775. I'll double-check how that got fixed,

fbrausse commented:
> #775. I'll double-check how that got fixed,

It's the deemed-too-hacky fix from #781. Should I remove it or shall we revisit this issue?

lucasccordeiro commented:
We could revisit it later.

fbrausse committed:
[cmake] don't depend on cheri-compressed-cap on non-CHERI builds

fbrausse committed:
[config] do not accept --cheri on non-CHERI builds

fbrausse committed:
[BUILDING] make Codacy (more) happy

fbrausse committed:
[util] c_typecast: remove assertion too strong for the C++ frontend

fbrausse committed:
[cmake] fix for building w/o cheri-compressed-cap

fbrausse committed:
[dereference] add typecast for operand of offset computation

fbrausse committed:
[regression] disable test failing on Debug build due to wchar_t literals

This test case employs wide character literals. The string_constantt class
assigns a wrong value for those, so the assertion in its set_value() triggers.
This makes the test case KNOWNBUG for Debug builds and vacuously successful
when assertions are disabled.

See also <https://github.com/esbmc/esbmc/issues/1165>.

fbrausse commented:
I've disabled `regression/esbmc-unix/github_1009_success` due to #1165. From the commit message:

This test case employs wide character literals. The string_constantt class
assigns a wrong value for those, so the assertion in its set_value() triggers.
This makes the test case KNOWNBUG for Debug builds and vacuously successful
when assertions are disabled.

**Edit:** This also affects
- regression/esbmc-unix2/github_615_[123]
- regression/esbmc/github_330_7

fbrausse committed:
[jimple] __ESBMC_alloc_size takes size_type() elements, not uint_type()

fbrausse committed:
[regression] disable more wchar_t cases, see #1165

fbrausse committed:
[regression] fixes the regression/esbmc/github_785 case

lucasccordeiro commented:
That's fine, @fbrausse.

lucasccordeiro commented:
@fbrausse: can you please rebase this PR with the master branch?

fbrausse committed:
[regression] make regression/esbmc/github_28_successful windows compatible

lucasccordeiro commented:
@fbrausse : can I merge this PR? :-)

fbrausse committed:
[regression] enable cheri tests conditional on cheri-clang

fbrausse committed:
[cheri] rm commented code

fbrausse committed:
[cheri] rm more commented code

fbrausse committed:
[clang-c] print exact clang invocation cmdline in Debug mode

fbrausse committed:
[cheri] clarify TODO

fbrausse committed:
[cheri] document TODO

fbrausse commented:
> @fbrausse: can you please rebase this PR with the master branch?

done.

fbrausse committed:
clang-format

lucasccordeiro commented:
Thanks for submitting this PR, @fbrausse!
